### PR TITLE
DOC-43: Documentation of GOTO model (de)serialisation.

### PR DESCRIPTION
**Pull request by marek-trtik**

### CODE REVIEW

File: src/goto-programs/README.md [hunk 1]
```diff
@@ -377,3 +377,181 @@ This stage concludes the *analysis-independent* program transformations.
...
+The serialisation is implemented in C++ modules:
+  - `write_goto_binary.h`
+  - `write_goto_binary.cpp`
+
+To serialise a `goto_modelt` instance `gm` to a stream `ostr` call the function `write_goto_binary`, e.g. `write_goto_binary(ostr, gm)`.
+
+The content of the written stream will have this strucutre:
+  - The header:
+    - A magic number: byte `0x7f` followed by 3 characters `GBF`.
+    - A version number written in the 7-bit encoding. Currently, only version `4` is supported.
```
Comment (tautschnig) [inline src/goto-programs/README.md]: Could a link explaining "7-bit encoding" be provided? It's used a lot below.

File: src/goto-programs/README.md [hunk 2]
```diff
@@ -377,3 +377,181 @@ This stage concludes the *analysis-independent* program transformations.
...
+      - The `irept` instance `s.type`.
+      - The `irept` instance `s.value`.
+      - The `irept` instance `s.location`.
+      - The string `s.name`.
+      - The string `s.module`.
+      - The string `s.base_name`.
+      - The string `s.mode`.
+      - The string `s.pretty_name`.
+      - The number `0` in the 7-bit encoding.
+      - The flags word in the 7-bit encoding. The bits in the flags word correspond to the following `bool` fields (from the most significant bit):
```
Comment (tautschnig) [inline src/goto-programs/README.md]: I'd use "Boolean" instead of "`bool`".

File: src/goto-programs/README.md [hunk 3]
```diff
@@ -377,3 +377,181 @@ This stage concludes the *analysis-independent* program transformations.
...
+        - `s.is_thread_local`
+        - `s.is_file_local`
+        - `s.is_extern`
+        - `s.is_volatile`
+  - The functions with bodies, i.e. those missing a body are skipped.
+    - The number of functions with bodies in the 7-bit encoding.
+    - The array of individual function with bodies. Each written function has this structure:
+      - The string with the name of the function.
+      - The number of instructions in the body of the function in the 7-bit encoding.
+      - The array of individual instruction in function's body. Each written instruction `I` has this structure:
```
Comment (tautschnig) [inline src/goto-programs/README.md]: individual instruction_s_

File: src/goto-programs/README.md [hunk 4]
```diff
@@ -377,3 +377,181 @@ This stage concludes the *analysis-independent* program transformations.
...
+        - The empty string (representing former `I.event`).
+        - The word in the 7-bit encoding `I.target_number`, i.e. the jump target to this instruction from other instructions.
+        - The word in the 7-bit encoding `I.targets.size()`, i.e. the count of jump targets from this instruction.
+        - The array of individual jump targets from this instruction, each written as a word in the 7-bit encoding.
+        - The word in the 7-bit encoding `I.labels.size()`.
+        - The array of individual labels, each written as a word in the 7-bit encoding.
+
+An important propery of the serialisation is that each serialised `irept`
+instance occurs in the stream exactly once. Namely, in the position of
+its first serialisation query. All other such queries save only hash
```
Comment (tautschnig) [inline src/goto-programs/README.md]: only _a_ hash

File: src/goto-programs/README.md [hunk 5]
```diff
@@ -377,3 +377,181 @@ This stage concludes the *analysis-independent* program transformations.
...
+call the function `read_goto_binary`, e.g.
+`read_goto_binary("/some/path/name.gbf", gm, message_handler)`, where
+`message_handler` must be an instance of `message_handlert` and serves
+for reporting issues during the process.
+
+The passed binary file is assumed to have the same structure as described in
+the [previous subsection](\ref subsection-goto-binary-serialisation).
+The process of the deserialisation does not involve any seeking in the file.
+The content is read linearly from the beginning to the end. `irept` instances
+and their string constats are deserialised into the memory only once at their
```
Comment (tautschnig) [inline src/goto-programs/README.md]: s/constats/constants/

File: src/goto-programs/README.md [hunk 6]
```diff
@@ -377,3 +377,181 @@ This stage concludes the *analysis-independent* program transformations.
...
+first occurreces in the stream. All subsequent deserialisation queries are
```
Comment (tautschnig) [inline src/goto-programs/README.md]: s/occurreces/occurrence/

File: src/goto-programs/README.md [hunk 7]
```diff
@@ -377,3 +377,181 @@ This stage concludes the *analysis-independent* program transformations.
...
+`goto-cc`, into an ELF image. Then the deserialisation has a support of
+automatic detection of that section in an ELF file and the deserialisation
+will be automatically started from that section.
+
+For reading the ELF images there is used an instance of `elf_readert`
+implemented in the C++ module:
+  - `elf_reader.h`
+  - `elf_reader.cpp`
+
+\subsubsection subsection-goto-binary-deserialisation-from-mac-o-fat-image Deserialisation from Mac-O fat image
```
Comment (tautschnig) [inline src/goto-programs/README.md]: It's "mach-o", not "mac-o" (multiple instances below as well)

File: src/util/README.md [hunk 8]
```diff
@@ -144,3 +144,51 @@ To be documented.
...
+This is implemented in C++ modules:
+  - `irep_serialization.h`
+  - `irep_serialization.cpp`
+  - `irep_hash_container.h` (applies only to `irept` instances)
+  - `irep_hash_container.cpp` (applies only to `irept` instances)
+
+\subsubsection irep-serialization-numbers Serialization of Numbers
+
+A number is serialiased in 7-bit encoding. For example. given a 2-byte
+number in base 2, like `10101010 01010101`, then it is save in 3 bytes,
```
Comment (tautschnig) [inline src/util/README.md]: is is save_d_

File: src/util/README.md [hunk 9]
```diff
@@ -144,3 +144,51 @@ To be documented.
...
+where each byte takes only 7 bits from the number, reading from the
+left. The 8th bit in each output byte is set to 1 except in the last
+byte, where the bit is 0. That 0 bit indicates the end of the end
```
Comment (tautschnig) [inline src/util/README.md]: duplicate "the end"

File: src/goto-programs/README.md [hunk 10]
```diff
@@ -377,3 +377,181 @@ This stage concludes the *analysis-independent* program transformations.
 \subsubsection goto-program-example-1-section Unsigned mult (unsigned a, unsigned b) { int acc, i; for (i = 0; i < b; i++) acc += a; return acc; }
 
 To be documented.
+
+
+\section section-goto-binary Binary Represenration
```
Comment (peterschrammel) [inline src/goto-programs/README.md]: Representation

File: src/goto-programs/README.md [hunk 11]
```diff
@@ -377,3 +377,181 @@ This stage concludes the *analysis-independent* program transformations.
...
+
+\subsection subsection-goto-binary-serialisation Serialisation
+
+The serialisation is implemented in C++ modules:
+  - `write_goto_binary.h`
+  - `write_goto_binary.cpp`
+
+To serialise a `goto_modelt` instance `gm` to a stream `ostr` call the function `write_goto_binary`, e.g. `write_goto_binary(ostr, gm)`.
+
+The content of the written stream will have this strucutre:
```
Comment (peterschrammel) [inline src/goto-programs/README.md]: structure

File: src/goto-programs/README.md [hunk 12]
```diff
@@ -377,3 +377,181 @@ This stage concludes the *analysis-independent* program transformations.
...
+        - The array of individual jump targets from this instruction, each written as a word in the 7-bit encoding.
+        - The word in the 7-bit encoding `I.labels.size()`.
+        - The array of individual labels, each written as a word in the 7-bit encoding.
+
+An important propery of the serialisation is that each serialised `irept`
+instance occurs in the stream exactly once. Namely, in the position of
+its first serialisation query. All other such queries save only hash
+code (i.e. reference) of the `irept` instance.
+
+The similar strategy is used for serialisation of string constants
```
Comment (peterschrammel) [inline src/goto-programs/README.md]: A similar

File: src/goto-programs/README.md [hunk 13]
```diff
@@ -377,3 +377,181 @@ This stage concludes the *analysis-independent* program transformations.
...
+shared amongst `irept` instances. Such a string is fully saved only in
+the first serialisation query of an `irept` instance it appears in and
+all other queries only save its integer has code.
```
Comment (peterschrammel) [inline src/goto-programs/README.md]: as code?

File: src/goto-programs/README.md [hunk 14]
```diff
@@ -377,3 +377,181 @@ This stage concludes the *analysis-independent* program transformations.
...
+
+\subsection subsection-goto-binary-deserialisation Deserialisation
+
+The deserialisation is implemented in C++ modules:
+  - `read_goto_binary.h`
+  - `read_goto_binary.cpp`
+  - `read_bin_goto_object.h`
+  - `read_bin_goto_object.cpp`
+
+First two modules are responsible for location of the stream with the
```
Comment (peterschrammel) [inline src/goto-programs/README.md]: The first two

File: src/goto-programs/README.md [hunk 15]
```diff
@@ -377,3 +377,181 @@ This stage concludes the *analysis-independent* program transformations.
...
+serialised data within a passed file. And the remaining two modules
+perform the actual deserialisation of `goto_modelt` instance from
```
Comment (peterschrammel) [inline src/goto-programs/README.md]: of a

File: src/goto-programs/README.md [hunk 16]
```diff
@@ -377,3 +377,181 @@ This stage concludes the *analysis-independent* program transformations.
...
+  - `osx_fat_reader.h`
+  - `osx_fat_reader.cpp`
+
+NOTE: This functionality is available only when the modules are built
+on a MacOS machine.
+
+\subsubsection subsection-goto-binary-is-binary-file Checking file type
+
+You can use function `is_goto_binary` to check whether a passed file contains
+a sederialised `goto_modelt` instance or not. This is done by checking the
```
Comment (peterschrammel) [inline src/goto-programs/README.md]: deserialised

File: src/util/README.md [hunk 17]
```diff
@@ -144,3 +144,51 @@ To be documented.
...
+The deserialisation does the oposite process and it is implemented in
+`irep_serializationt::read_gb_word`.
+
+\subsubsection irep-serialization-strings Serialization of Strings
+
+A string is encoded as classic 0-terminated C string. However,
+characters `0` and `\\` are escaped by writing additional `\\` 
+before them. 
+
+This is implmented in the function `write_gb_string` and the
```
Comment (peterschrammel) [inline src/util/README.md]: Prefix function with class name so that it is linked by doxygen

**Comment by marek-trtik** [comment]
Requested updates are available.

**Review by peterschrammel — APPROVED** [review]

**Comment by peterschrammel** [comment]
@marek-trtik, please rebase.

**Comment by marek-trtik** [comment]
@peterschrammel Done.

**Comment by tautschnig** [comment]
@marek-trtik Would you mind squashing your commits? Then this one should be ready to be merged.

**Comment by marek-trtik** [comment]
@tautschnig Done.

**Review by allredj — COMMENTED** [review]
This PR failed Diffblue compatibility checks (cbmc commit: 6cd614b6b759d0a5c898a0cb005e22eeab65564e).
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

**Pull request merged**